semgrep-core/src/fixing/Autofix_printer.ml

```diff
@@ -0,0 +1,148 @@
+ *)
+(******************************************************************************)
+
+type ast_node_source =
+  (* Indicates that a node came from the target file via a metavar binding *)
+  | Target
+  (* Indicates that a node came from the fix pattern *)
+  | FixPattern
+
+(* So that we can print by lifting the original source for unchanged AST nodes,
+ * this indicates whether a given AST node came from the target file (via
+ * metavariable bindings) or from the rule's fix pattern. Nodes that do not
+ * appear in this table came from neither. *)
+type ast_node_table = (AST_generic.any, ast_node_source) Hashtbl.t
+
+module PythonPrinter = Hybrid_print.Make (struct
+  class printer = Ugly_print_AST.python_printer
+end)
+
+let get_printer lang external_printer : Ugly_print_AST.printer_t option =
+  match lang with
+  | Lang.Python -> Some (new PythonPrinter.printer external_printer)
+  | _ ->
+      logger#info "Failed to render autofix: no printer available for %s"
+        (Lang.to_string lang);
+      None
+
+let original_source_of_ast source any =
+  let* start, end_ = Visitor_AST.range_of_any_opt any in
+  let starti = start.Parse_info.charpos in
+  let _, _, endi = Parse_info.get_token_end_info end_ in
+  let len = endi - starti in
+  let str = String.sub source starti len in
+  Some str
+
+(* Add each metavariable value to the lookup table so that it can be identified
+ * during printing *)
+let add_metavars (tbl : ast_node_table) metavars =
+  List.iter
+    (fun (_, mval) ->
+      let any = MV.mvalue_to_any mval in
+      Hashtbl.replace tbl any Target;
+      (* For each metavariable binding that is a list of things, we need to
+       * iterate through and add each item in the list to the table as well.
+       *
+       * For example, if `$...BAR` is bound to the separate arguments `1` and
+       * `2`, in the example below, then the complete argument list `(1, 2)`
+       * would never appear in the resulting AST that we attempt to print.
+       * Despite that, we would like to reuse the original text for `1` and `2`.
+       *
+       * foo($...BAR, 5) -> foo(1, 2, 5)
+       *
+       * We don't need to recurse any deeper, because individual list items are
+       * the smallest components of a metavariable value that will be lifted
+       * into the resulting AST.
+       * *)
+      match mval with
+      | MV.Args args ->
+          List.iter (fun arg -> Hashtbl.replace tbl (Ar arg) Target) args
+      (* TODO iterate through other metavariable values that are lists *)
+      | _ -> ())
+    metavars
+
+(* Add each AST node from the fix pattern AST to the lookup table so that it can
+ * be identified during printing *)
+let add_fix_pattern_ast_nodes (tbl : ast_node_table) ast =
+  let visitor =
+    Visitor_AST.(
+      mk_visitor
+        {
+          default_visitor with
+          kargument =
+            (fun (k, _) arg ->
+              Hashtbl.replace tbl (Ar arg) FixPattern;
```

Comment:
I believe that what would happen right now is that we would fail to replace the `$X` because ordinary expressions are not yet handled in the metavariable replacement step. Then, when checking the resulting AST for lingering metavariables, we would find `$X` and abort at that point.

Once we implement metavariable replacement for this case (which won't be hard, I just wanted to get one scenario working end-to-end first), then we'd replace `$X` with, say, `5`. We'd still add the AST node for `Arg (1 + $X)` here (pardon the pseudocode), but when printing, we'd look up `Arg (1 + 5)`. Since those two nodes are not equal, the hashtbl lookup will fail and we will not pull the original text from the fix pattern.